An experimental SPIR-V backend for LLVM based on GlobalISel.

mantognini opened the pull request:
While this is a work-in-progress, this backend already supports a significant subset of LLVM IR as input. Refer to the test suite to get an idea about the coverage.

Additional details are available in the documentation (llvm/docs/SPIR-V-Backend.rst).

Commit:
[SPIR-V] Add a prototype SPIR-V backend

This new, experimental backend is based on GlobalISel. Original work
was done by Lewis Crawford during a summer internship at ARM in 2019.

mantognini committed:
[SPIR-V] Add documentation for SPIR-V Backend

This documentation is intended as an entry point for future
collaborators, listing build instructions, where the main components of
the backend are and what areas can be improved.

mantognini committed:
[SPIR-V] Update README.md with a link to the backend documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/KhronosGroup/LLVM-SPIRV-Backend?pullRequest=1) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/KhronosGroup/LLVM-SPIRV-Backend?pullRequest=1) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/KhronosGroup/LLVM-SPIRV-Backend?pullRequest=1) it.</sub>

DiamondLovesYou commented:
This is amazing work! I maintain a [Rust fork](https://github.com/geobacter-rs) which supports single source accelerator programming by loading Rust's own compiler metadata and re-codegen-ing functions for new targets at runtime. It *used* to kinda support targeting Vulkan's SPIRV (nVidia killed OpenCL, sorry..), but I dropped support to work on a target which didn't require so much work to make work reliably, that target being AMDGPU.

Anyway, I have a few comments coming from the Rust world and targeting the Vulkan world. I've had some experience modifying the "original" translator ([available here](https://github.com/geobacter-rs/SPIRV-LLVM-Translator)) in order to make it suitable for Rust LLVM IR and Vulkan; while looking over this PR I've noticed that as is this will have some of the same issues I had with that. At any rate, I don't think these comments should block; I mean these points to be informative.

Some issues that will be problematic for Rust:
* Rust uses zero sized arrays and structures in a couple of ways: alignment, occasionally padding, and sometimes to represent dynamically sized types. Rust will insert a zero sized array between nearly every structure member. Empty structures in ordinary code is common as well, but not IIRC used as a tool by the compiler like arrays.
Here is an example:
`%"core::fmt::Arguments" = type { [0 x i64], { [0 x { [0 x i8]*, i64 }]*, i64 }, [0 x i64], { i64*, i64 }, [0 x i64], { [0 x { i8*, i8* }]*, i64 }, [0 x i64] }`
Of course, these are invalid in SPIRV. The obvious solution was to "just" rewrite GEP/extractvalue/insertvalue/etc to drop those types/indices, while respecting the alignment they force for the next member element by inserting decorations or whatnot.
* Use of mangled names is/was always weird:
  * Functions: they should be turned into intrinsics. Using mangled link names here is possible in today's Rust, so this isn't a huge deal if *really* ugly.
  * Special opaque types (like images, pipes): today's Rust can't generate these. For one, empty structs produce an empty LLVM struct type, and not an opaque type. And second, the period in `opencl.*` means no empty struct would have a valid identifier. So, this will probably require a eg `#[repr(opaque = "..")]` attribute to support, if nothing else.
* Address space cast optimization. These casts aren't normally handled at all by Rust, at least upstream. But I've written a patch which adds the capability for it. In order to keep the patch small however, the patch can be kinda conservative and cast pointers to the flat address space (borrowed from AMDGPU, meaning the generic address space here) when it doesn't need to. As a result I've found LLVM's optimizations in this area to be fairly weak.
Granted, the `Kernel` cap is required for casts to the generic space (which Rust requires to support function calls with pointer parameters, or just pointer values in general). AMDVLK was nice enough to ignore this distinction when it would parse and translate SPIRV back into LLVM IR, so it still worked. Mesa, OTOH, would segfault...
* ConstantExpr address space casts. This was a problem for AMDGPU; I don't think this is handled here either.
* Function metadata for SPIRV entry point level attributes (Exec mode, Exec model, etc).
* Decorations. I used a crude metadata spec get the proper decorations emitted for Vulkan SPIRV. Of course, SPIRV <-> LLVM is pretty awkward here: many Vk required decorations need to be on types and LLVM doesn't allow types to have metadata... Using global variable metadata for this is pretty awful, in particular for Rust due to the extra types everywhere, you have to march in lock step through the metadata spec tree and the type tree (while disregarding types that would be discarded!) of the global to construct the SPIRV type with all the proper decorations. Pretty error prone.
* Global metadata to mark builtins/etc (Vk specific, of course).

As I said, Geobacter doesn't support SPIRV out of the box any longer (but that might change soon thanks to you!), so at best I'm just a passerby. Anyway, this is exciting work! :+1:

mantognini commented on llvm/docs/SPIR-V-Backend.rst:
```diff
@@ -0,0 +1,294 @@
+
+-  SPIRVAddRequirements::runOnMachineFunction
+
+   SPIRVAddRequirements iterates over all instructions and inserts
```
The first part should already be implemented, but I agree the doc/comments should be updated to clarify that point. The configurability-part is currently listed as a TODO. I'll prepare a commit to update the doc soonish, and also add a test case.

Mind the fact that currently only `SPV_KHR_no_integer_wrap_decoration` is enabled: 
```cpp
// TODO use command line args for this rather than defaults
void SPIRVSubtarget::initAvailableExtensions(const Triple &TT) {
  using namespace Extension;
  if (TT.isVulkanEnvironment()) {
    availableExtensions = {};
  } else {
    // A default extension for testing - should use command line args
    availableExtensions = {SPV_KHR_no_integer_wrap_decoration};
  }
}
```
https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-9ac8724b38f0f2f16078e93ad5448a6fR151

AlexeySachkov replied:
Thanks! BTW, what about restricting target SPIR-V version? I.e. is there any infrastructure to generate, for example, SPIR-V 1.2 compatible output file?

mantognini replied:
We do have min/max version information about capabilities, decorations, etc... [1] that is used to ensure the output is valid, but currently we always assume SPIR-V 1.4 and there's a TODO to have a command line option, or something, to chose the desired SPIR-V version. [2] Once this TODO is implemented we should also add the expected tests, of course.

1. https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-3253aa7cd0b28412077e3f870178d895R105-R129
2. https://github.com/KhronosGroup/LLVM-SPIRV-Backend/pull/1/files#diff-9ac8724b38f0f2f16078e93ad5448a6fR45-R47

AlexeySachkov replied:
Thanks!

mantognini commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -0,0 +1,1334 @@
+  case SyncScope::System:
+    return Scope::Device;
+  default:
+    // TODO is there any other way that more scopes could be defined?
```
Thanks for the clarification.

mantognini reviewed (commented):
@DiamondLovesYou, thanks for your comment. There are definitely valuable experience to be taken from that. I'm not that familiar with Rust and its intrinsic challenges, and have no specific experience yet with Vulkan. I guess what would be greatly beneficial at this stage, and even later, are some test cases. Do you think you would be able to write some that illustrate the points you made? Hopefully, some of your points are not specific to this backend and could be addressed in LLVM directly.

BTW, after this PR is merged, I'd like to rebase the backend on a more recent LLVM version. For example, there were a few changes to GlobalISel that require a few minor patches.

mantognini committed:
[SPIR-V] Clarify code and doc related to SPIRVAddRequirementsPass

The SPIRVAddRequirements inserts the appropriate OpCapability and
OpExtension instructions. The code and documentation are clarified in
that respect. A test case is also added, showing that the backend
support the SPV_KHR_no_integer_wrap_decoration extension.

The code is also simplified in a few places to use references when
pointers are never null and therefore not needed.

mantognini committed:
[SPIR-V] Clarify SyncScope handling

Additional scope can be added using LLVMContext::getOrInsertSyncScopeID,
but this backend doesn't handle any target-scope at this stage.

neildhickey commented:
Approved
